### Importing libraries


In [27]:
import pandas as pd
import numpy as np


### Loading the data


In [28]:
df=pd.read_csv(r"F:\Data Analytics\MElbourne\melbourne.csv")
df.head(5)

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode      ...       Bathroom  Car  Landsize  \
0  03-09-2016       2.5    3067.0      ...            1.0  1.0     126.0   
1  03-12-2016       2.5    3067.0      ...            1.0  1.0     202.0   
2  04-02-2016       2.5    3067.0      ...            1.0  0.0     156.0   
3  04-02-2016       2.5    3067.0      ...            2.0  1.0       0.0   
4  04-03-2017       2.5    3067.0      ...            2.0  0.0     134.0   

   BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0           NaN        NaN        Yarra  -37.8014    144.9958   
1           NaN        NaN        Yarra  -37.7996    144.9984   
2          79.0     1900.0        Yarra  -37.8079    144.9934   
3           NaN        NaN        Yarra  -37.8114    145.0116   
4         150.0     1900.0        Yarra  -37.8093    144.9944   

              Regionname Propertycount  
0  Northern Metropolitan        4019.0  
1  Northern Metropolitan        4019.0  
2  Northern Metropolitan        4019.0  
3  Northern Metropolitan        4019.0  
4  Northern Metropolitan        4019.0  

[5 rows x 21 columns]

In [29]:
#checking the info of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
Suburb           23547 non-null object
Address          23547 non-null object
Rooms            23547 non-null int64
Type             23547 non-null object
Price            18396 non-null float64
Method           23547 non-null object
SellerG          23547 non-null object
Date             23547 non-null object
Distance         23546 non-null float64
Postcode         23546 non-null float64
Bedroom2         19066 non-null float64
Bathroom         19063 non-null float64
Car              18921 non-null float64
Landsize         17410 non-null float64
BuildingArea     10018 non-null float64
YearBuilt        11540 non-null float64
CouncilArea      15656 non-null object
Lattitude        19243 non-null float64
Longtitude       19243 non-null float64
Regionname       23546 non-null object
Propertycount    23546 non-null float64
dtypes: float64(12), int64(1), object(8)
memory usage: 3.8+ M

In [30]:
df.shape

(23547, 21)

In [31]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

#### checking the null values in the form of %ge


In [32]:
round(((100*df.isnull().sum())/len(df.index)),2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

In [33]:
df.shape

(23547, 21)

In [34]:
#Dropping the columns (BuildingArea, YearBui1t, CouncilArea) as their missing data is quite high(more than 30%)
#df=df.drop(['BuildingArea','YearBuilt','CouncilArea'],axis=1)
df = df.drop(['BuildingArea','YearBuilt','CouncilArea'],axis=1)

In [35]:
print(df.shape)

(23547, 18)


In [36]:
#Now we will analyze the empty data in the rows. For that we will first remove the rows having the missing value greater than
#a particular threshold value(5 in this case)
#sum is from numpy therefor axis=1 (for rows)
df=df[df.isnull().sum(axis=1)<=5]
df.shape

(19269, 18)

Now focussing on the Price column ,
We will remove the columns having the null value in the price column


In [37]:
df=df[~np.isnan(df.Price)]

In [38]:
df.shape
round(((100*df.isnull().sum())/len(df.index)),2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.76
Landsize         9.83
Lattitude        0.15
Longtitude       0.15
Regionname       0.00
Propertycount    0.00
dtype: float64

In [39]:
df.loc[:,('Bedroom2','Bathroom','Car','Landsize','Lattitude','Longtitude')].describe()

Bedroom2      Bathroom           Car       Landsize     Lattitude  \
count  14927.000000  14925.000000  14820.000000   13603.000000  15064.000000   
mean       2.913043      1.538492      1.615520     558.116371    -37.809849   
std        0.964641      0.689311      0.955916    3987.326586      0.081152   
min        0.000000      0.000000      0.000000       0.000000    -38.182550   
25%        2.000000      1.000000      1.000000     176.500000    -37.858100   
50%        3.000000      1.000000      2.000000     440.000000    -37.803625   
75%        3.000000      2.000000      2.000000     651.000000    -37.756270   
max       20.000000      8.000000     10.000000  433014.000000    -37.408530   

         Longtitude  
count  15064.000000  
mean     144.996338  
std        0.106375  
min      144.431810  
25%      144.931193  
50%      145.000920  
75%      145.060000  
max      145.526350

In [40]:
# We find that std of landsize column is very high so removing the nan values of the Landsize column
df=df[~np.isnan(df.Landsize)]


In [41]:
df.shape

(13603, 18)

In [42]:
df.loc[:,('Bedroom2','Bathroom','Car','Lattitude','Longtitude')].describe()

Bedroom2      Bathroom           Car     Lattitude    Longtitude
count  13603.000000  13602.000000  13540.000000  13581.000000  13581.000000
mean       2.914945      1.534921      1.610414    -37.809204    144.995221
std        0.966145      0.691834      0.962244      0.079257      0.103913
min        0.000000      0.000000      0.000000    -38.182550    144.431810
25%        2.000000      1.000000      1.000000    -37.856820    144.929600
50%        3.000000      1.000000      2.000000    -37.802360    145.000100
75%        3.000000      2.000000      2.000000    -37.756400    145.058320
max       20.000000      8.000000     10.000000    -37.408530    145.526350

In [43]:
## replacing the missing valeus with mean in the lattitud and longitude column
lat_mean=np.mean(df.Lattitude)
lon_mean=np.mean(df.Longtitude)
df.Lattitude=df.Lattitude.replace(np.NAN,value=lat_mean)
df.Longtitude=df.Longtitude.replace(np.NAN,value=lon_mean)

In [44]:
round(((100*df.isnull().sum())/len(df.index)),2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.00
Longtitude       0.00
Regionname       0.00
Propertycount    0.00
dtype: float64

In [45]:
#we will fill the remaini columns (Bathrooms and Cars with the mode value)
df.Bathroom.fillna(value=df.Bathroom.mode,inplace=True)

In [46]:
df.Car.fillna(value=df.Car.mode,inplace=True)

In [47]:
df.shape

(13603, 18)

In [48]:
df.isnull().sum()
#df[np.isnan(df.Bathroom)].fillna(value=1.0)

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

## In this way we have cleaned our data and the final dimension of dataframe is 13603 row and 18 coumns